pkgs/development/tools/misc/elfutils/default.nix

```diff
@@ -77,15 +77,43 @@ stdenv.mkDerivation rec {
 
   enableParallelBuilding = true;
 
+  postFixup = lib.optionalString enableDebuginfod ''
+    moveToOutput etc $libdebuginfod
+
+    # move libdebuginfod.so to a different output so that only stuff that
+    # uses libdebuginfod.so transitively depends on curl and libarchive.
+    mkdir -p $libdebuginfod/lib
+    for i in $out/lib/libdebuginfod*; do
+      mv -v $i $libdebuginfod/lib/
+    done
+
+    # fix the executables depending on it
+    for i in $bin/bin/debuginfod*; do
+      origRpath="$(patchelf --print-rpath "$i")"
+      patchelf --set-rpath "$libdebuginfod/lib:$origRpath" "$i"
+    done
```

Comment:
The need for `patchelf` to keep the package running always irks me: it always has a potential missing some reference to patch. Why not keep it in `$lib` instead? `elfutils` is a toolchain-style package I would say. It's OK to have a few extra dependencies. Looking at the individual deps you mention:

- `libarchive` seems to pull in extra 2MB: 800K of libarchive itself and 1.2MB of libxml2 (the rest looks like baseline packages)
- `libcurl` seems to pull in extra 5.2MB: curl itself (800KB), libktb5 (2.2MB), brotli (1.7MB) (the rest looks like baseline packages)

I would say these are benign and worth pulling in unconditionally. The `curl` one is especially weak I would say as `curl` is already present in a typical system.

WDYT of just installing `debuginfod` by default?

Reply:
> WDYT of just installing debuginfod by default?

tbh I expected that people would oppose that, given that elfutils is in the closure of approximately everything through systemd (udev), hence the extra work I did here. If you don't mind I'll wait for a second opinion before going in this direction.

> The need for patchelf to keep the package running always irks me: it always has a potential missing some reference to patch.

there are install tests, and they catch the regression if one remove the patchelf invocation, so that seems unlikely.



To give actual numbers: elfutils.bin grows from 47MB to 73MB in closure size by enabling debuginfod support, and the exact list of added dependencies is sqlite, curl, libmicrohttpd, libarchive. It is true that systemd depends on curl and libmicrohttpd, and that nix depends the remaining two, so we will likely have them in most nixos closures.